control-plane/agents/core/src/pool/specs.rs

```diff
@@ -429,51 +443,93 @@ impl ResourceSpecsLocked {
-            } else if !spec.state.created() {
-                return Err(SvcError::ReplicaNotFound {
-                    replica_id: request.uuid.clone(),
-                });
-            } else if spec.share == Protocol::Off {
-                return Err(SvcError::NotShared {
-                    kind: ResourceKind::Replica,
-                    id: request.uuid.to_string(),
-                });
+            let spec_clone = {
+                let status = registry.get_replica(&request.uuid).await?;
+                let mut spec = replica_spec.lock().await;
+                if spec.pending_op() {
+                    return Err(SvcError::DirtySpec {
+                        kind: ResourceKind::Replica,
+                        id: request.uuid.to_string(),
+                    });
+                } else if spec.updating {
+                    return Err(SvcError::Conflict {});
+                } else if !spec.state.created() {
+                    return Err(SvcError::ReplicaNotFound {
+                        replica_id: request.uuid.clone(),
+                    });
+                } else if spec.share == Protocol::Off
+                    && status.share == Protocol::Off
+                {
+                    return Err(SvcError::NotShared {
+                        kind: ResourceKind::Replica,
+                        id: request.uuid.to_string(),
+                    });
+                }
+
+                spec.updating = true;
+                spec.start_op(ReplicaOperation::Unshare);
+                spec.clone()
+            };
+
+            if let Err(error) = registry.store_obj(&spec_clone).await {
+                let mut spec = replica_spec.lock().await;
+                spec.updating = false;
+                spec.clear_op();
+                return Err(error);
             }
 
-            spec.updating = true;
-            let mut spec_clone = spec.clone();
-            drop(spec);
+            let result = node.unshare_replica(request).await;
+            Self::replica_complete_op(
+                registry,
+                result,
+                replica_spec,
+                spec_clone,
+            )
+            .await
+        } else {
+            node.unshare_replica(request).await
+        }
+    }
 
-            match node.unshare_replica(request).await {
-                Ok(_) => {
-                    spec_clone.share = Protocol::Off;
-                    let result = {
-                        let mut store = registry.store.lock().await;
-                        store.put_obj(&spec_clone).await
-                    };
-                    if let Err(error) = result {
-                        let _ =
-                            node.share_replica(&request.clone().into()).await;
-                        let mut spec = replica_spec.lock().await;
-                        spec.updating = false;
-                        return Err(error.into());
+    async fn replica_complete_op<T>(
```

Comment:
Some comments in this function would be good as it will give it context.

Reply:
Done